identity/app/controllers/ChangePasswordController.scala

```diff
@@ -79,44 +80,47 @@ class ChangePasswordController(
   def renderPasswordConfirmation(returnUrl: Option[String]): Action[AnyContent] = Action { implicit request =>
     val idRequest = idRequestParser(request)
     val userIsLoggedIn = authenticationService.userIsFullyAuthenticated(request)
-    NoCache(DiscardingIdentityCookies(
+    NoCache(
       Ok(
         IdentityHtmlPage.html(
           views.html.password.passwordResetConfirmation(page, idRequest, idUrlBuilder, userIsLoggedIn, returnUrl, None)
         )(page, request, context)
-      )
-    ))
+      ))
   }
 
   def submitForm(): Action[AnyContent] = csrfCheck {
     fullAuthAction.async {
       implicit request =>
         val idRequest = idRequestParser(request)
         val boundForm = passwordFormWithConstraints.bindFromRequest()
-        val futureFormOpt = boundForm.value map {
-          data =>
-            val update = PasswordUpdate(data.oldPassword, data.newPassword1)
-            api.updatePassword(update, request.user.auth, idRequest.trackingData) map {
-              case Left(errors) =>
-                boundForm.withError("oldPassword", Messages("error.wrongPassword"))
-
-              case Right(result) =>
-                boundForm
-            }
-        }
 
-        val futureForms = futureFormOpt getOrElse Future.successful(boundForm)
-        futureForms map { form =>
-          if(form.hasErrors){
-            val pwdExists = request.getQueryString("passwordExists") exists { _.toBoolean }
+        def onError(formWithErrors: Form[PasswordFormData]): Future[Result] = {
+          logger.info("form errors in change password attempt")
+          Future.successful(
             NoCache(
-              SeeOther(routes.ChangePasswordController.displayForm().url).flashing(clearPasswords(form).toFlash)
+              SeeOther(routes.ChangePasswordController.displayForm().url).flashing(clearPasswords(formWithErrors).toFlash)
             )
-          } else {
-            val userIsLoggedIn = authenticationService.userIsFullyAuthenticated(request)
-            NoCache(SeeOther(routes.ChangePasswordController.renderPasswordConfirmation(None).url))
-          }
+          )
         }
+
+        def onSuccess(form: PasswordFormData): Future[Result] = {
+            val update = PasswordUpdate(form.oldPassword, form.newPassword1)
+            val authResponse = api.updatePassword(update, request.user.auth, idRequest.trackingData)
+
+            signInService.getCookies(authResponse, true) map {
```

Comment:
Easier to read using named arguments for literals i.e. `rememberMe = true`

Reply:
Updated thanks